### PR TITLE
Lstm fixes

### PR DESCRIPTION
**Pull request by abergeron**
This fixes the memory leak in the DLT LSTM example and allows training to complete in float32 with the new backend.

### CODE REVIEW

File: theano/sandbox/gpuarray/blas.py [hunk 1]
```diff
@@ -239,6 +239,7 @@ def c_code(self, node, name, inputs, outputs, sub):
         dims[0] = PyGpuArray_DIMS(%(A)s)[0];
         dims[1] = PyGpuArray_DIMS(%(B)s)[1];
 
+        Py_XDECREF(%(out)s);
```
Comment (nouiz) [inline theano/sandbox/gpuarray/blas.py]: make an issue to tell that this op do not reuse preallocated output. With scan, we start to do that more frequently.

Ideally, we should only allocate if no output or output wrong shape.

File: theano/sandbox/gpuarray/type.py [hunk 2]
```diff
@@ -50,13 +55,20 @@ def filter(self, data, strict=False, allow_downcast=None):
...
+        elif (allow_downcast or
+              (allow_downcast is None and
+               type(data) == float and
+               self.dtype == config.floatX)):
             data = gpuarray.array(data, dtype=self.typecode, copy=False,
                                   ndmin=len(self.broadcastable))
         else:
+            if not hasattr(data, 'dtype'):
+                data = gpuarray.array(data, copy=False)
+
```
Comment (nouiz) [inline theano/sandbox/gpuarray/type.py]: why this extra call? If useful, add a comment that tell why.

File: theano/sandbox/gpuarray/blas.py (hunk 1 above)
Comment (nouiz) [inline theano/sandbox/gpuarray/blas.py]: In fact, it is so fast to implement it, can you do it instead of making a ticket?

**Comment by abergeron** [comment]
I've made the change to GpuDot22 and I've also made a similar change to all the other blas ops so that they attempt to reuse the output buffer in some cases.

File: theano/sandbox/gpuarray/blas.py [hunk 3]
```diff
@@ -262,19 +283,18 @@ def c_code(self, node, name, inputs, outputs, sub):
         return code
 
     def c_code_cache_version(self):
-        return (1,)
+        return (2,)
+
+    def c_header_dir(self):
+        ret = super(GpuDot22, self).c_header_dirs()
```
Comment (nouiz) [inline theano/sandbox/gpuarray/blas.py]: This isn't good. It can raise MethodNotImplemented and loose the new information.

Comment (abergeron) [inline theano/sandbox/gpuarray/blas.py]: No, because it inherits from BlasOp which does define c_header_dirs()

**Comment by abergeron** [comment]
I fixed all the problems.

File: theano/sandbox/gpuarray/blas.py [hunk 4]
```diff
@@ -262,19 +284,14 @@ def c_code(self, node, name, inputs, outputs, sub):
         return code
 
     def c_code_cache_version(self):
-        return (1,)
+        return (3,)
 
     def c_headers(self):
         ret = super(GpuDot22, self).c_headers()
-        return ret + ['<numpy_compat.h>']
+        return ret + ['<numpy_compat.h>', '"gpuarray_helper.h"']
```
Comment (nouiz) [inline theano/sandbox/gpuarray/blas.py]: Now this change isn't needed anymore. Also, numpy_compat is now added by blas op, so it isn't useful now, probably old stuff. So you can remove this method completly I think.

Comment (abergeron) [inline theano/sandbox/gpuarray/blas.py]: Fixed.

**Comment by nouiz** [comment]
The tests failing in gpuarray where already there, so I merge.

**Pull request merged**